x/pairing/keeper/msg_server_relay_payment_test.go

```diff
@@ -1300,3 +1348,85 @@ func TestCuUsageInProjectsAndSubscription(t *testing.T) {
+	require.Nil(t, err)
+
+	QoS := &types.QualityOfServiceReport{Latency: sdk.NewDecWithPrec(1, 0), Availability: sdk.NewDecWithPrec(1, 0), Sync: sdk.NewDecWithPrec(1, 0)}
+
+	relaySession := common.BuildRelayRequest(ts.ctx, ts.providers[0].Addr.String(), []byte(ts.spec.Apis[0].Name), ts.spec.Apis[0].ComputeUnits, ts.spec.Name, QoS)
+
+	currentEpoch := ts.keepers.Epochstorage.GetEpochStart(sdk.UnwrapSDKContext(ts.ctx))
+
+	// make the context's lava chain ID a non-empty string (as usually defined in most tests)
+	lavaChainID := "lavanet"
+	_ctx := sdk.UnwrapSDKContext(ts.ctx)
+	_ctxBlockHeader := _ctx.BlockHeader()
+	_ctxBlockHeader.ChainID = lavaChainID
+	_ctx = _ctx.WithBlockHeader(_ctxBlockHeader)
+	ts.ctx = sdk.WrapSDKContext(_ctx)
+
+	tests := []struct {
+		name         string
+		badgeAddress sdk.AccAddress // should be the badge user (in happy flow)
+		relaySigner  common.Account // should be the badge user (in happy flow)
+		badgeSigner  common.Account // should be the badge granter, i.e. project developer (in happy flow)
+		epoch        uint64
+		lavaChainID  string
+		cuAllocation uint64
+		valid        bool
+	}{
+		{"happy flow", badgeUser.Addr, badgeUser, projectDeveloper, currentEpoch, lavaChainID, ts.spec.Apis[0].ComputeUnits + 1, true},
+		{"badge address != badge user", projectDeveloper.Addr, badgeUser, projectDeveloper, currentEpoch, lavaChainID, ts.spec.Apis[0].ComputeUnits + 1, false},
+		{"relay signer != badge user", badgeUser.Addr, projectDeveloper, projectDeveloper, currentEpoch, lavaChainID, ts.spec.Apis[0].ComputeUnits + 1, false},
+		{"badge signer != project developer", badgeUser.Addr, badgeUser, badgeUser, currentEpoch, lavaChainID, ts.spec.Apis[0].ComputeUnits + 1, false},
+		{"badge epoch != relay epoch", badgeUser.Addr, badgeUser, projectDeveloper, currentEpoch - 1, lavaChainID, ts.spec.Apis[0].ComputeUnits + 1, false},
+		{"badge lava chain id != relay lava chain id", badgeUser.Addr, badgeUser, projectDeveloper, currentEpoch, "dummy-lavanet", ts.spec.Apis[0].ComputeUnits + 1, false},
+		{"badge cu allocation < relay cu sum", badgeUser.Addr, badgeUser, projectDeveloper, currentEpoch, lavaChainID, ts.spec.Apis[0].ComputeUnits - 1, false},
+		{"badge epoch != relay epoch (epoch passed)", badgeUser.Addr, badgeUser, projectDeveloper, currentEpoch, lavaChainID, ts.spec.Apis[0].ComputeUnits + 1, true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// this test was put to make sure that we only compare the relay's and badge's epoch fields (without any
+			// relation to the advancement of blocks). So this test should have no errors
+			if tt.name == "badge epoch != relay epoch (epoch passed)" {
+				ts.ctx = testkeeper.AdvanceEpoch(ts.ctx, ts.keepers)
+				ts.ctx = testkeeper.AdvanceBlock(ts.ctx, ts.keepers)
+
+				// remove past payments to avoid double spending error (first test had a successful payment)
```

Comment:
we test double spending in another unit test. There is no need to test it again here. Moreover, I don't want double spending to affect the test since it's a valid case and I want to test it individually (the case should not be influenced by other cases)